File: C# Fundamentals/C# OOP Basics/13. Polymorphism/Ex/P01_Vehicles/Models/Vehicle.cs
using System;

public abstract class Vehicle : IVehicle
{
    protected const double CarConsumptionIncrease = 0.9d;
    protected const double TruckConsumptionIncrease = 1.6d;
    protected const double BusConsumptionIncrease = 1.4d;
    protected const double TruckFuelLeak = 0.95d;

    protected const string FuelMustBePositiveMessage = "Fuel must be a positive number";
    protected const string FuelIsMoreThanCapacityMessage = "Cannot fit {0} fuel in the tank";

    public double FuelQuantity { get; set; }
    public double FuelConsumptionPerKm { get; set; }
    public double FuelTankCapacity { get; set; }

    public virtual string DriveEmpty(double kilometres) { return string.Empty; }
    public virtual void Refuel(double liters)
    {
        if (liters <= 0)
        {
            throw new InvalidOperationException(FuelMustBePositiveMessage);
        }
        if (this.FuelQuantity + liters > this.FuelTankCapacity)
        {
            throw new InvalidOperationException(string.Format(FuelIsMoreThanCapacityMessage, liters));
        }
        this.FuelQuantity += liters;
    }

    public Vehicle(double fuelQuant, double fuelConsum, double tankCapacity)
    {
        this.FuelQuantity = fuelQuant;
        this.FuelConsumptionPerKm = fuelConsum;
        this.FuelTankCapacity = tankCapacity;
        if (this.FuelQuantity > this.FuelTankCapacity)
        {
            this.FuelQuantity = 0;
        }
    }

    public virtual string Drive(double kilometres)
    {
        if(this.FuelQuantity <= this.FuelTankCapacity)
        {
            double neededFuel = kilometres * this.FuelConsumptionPerKm;
            if (this.FuelQuantity >= neededFuel)
            {
                this.FuelQuantity -= neededFuel;
                return $"{this.GetType()} travelled {kilometres} km";
            }
            else
            {
                return $"{this.GetType()} needs refueling";
            }
        }
        else
        {
            return string.Empty;
        }
    }
    public override string ToString()
    {
        return $"{this.GetType()}: {this.FuelQuantity:F2}";
    }
}

File: Java/01.Advanced/02.OOP/11.Polymorphism/Ex/src/t_02_vehicles_extend/Vehicle.java
package t_02_vehicles_extend;

public abstract class Vehicle {
    private double fuelQuantity;
    private double fuelConsumption;
    private double tankCapacity;

    private double addedFuelConsumption;

    public Vehicle(double fuelQuantity, double fuelConsumption, double tankCapacity) {
        this.setTankCapacity(tankCapacity);
        this.setFuelQuantity(fuelQuantity);
        this.setFuelConsumption(fuelConsumption);
    }

    public boolean drive(double distance, boolean airConditionerOn) {
        double currentFuelConsumption = fuelConsumption;

        if (airConditionerOn) {
            currentFuelConsumption += addedFuelConsumption;
        }

        if ((currentFuelConsumption * distance) <= fuelQuantity) {
            this.setFuelQuantity(fuelQuantity - (currentFuelConsumption * distance));
            return true;
        }

        return false;
    }

    public void refuel(double fuel) {
        if (fuel <= 0) {
            throw new IllegalArgumentException("Fuel must be a positive number");
        }
        this.setFuelQuantity(fuelQuantity + fuel);
    }

    protected void setAddedFuelConsumption(double addedFuelConsumption) {
        this.addedFuelConsumption = addedFuelConsumption;
    }

    public void setFuelQuantity(double fuelQuantity) {
        if (fuelQuantity < 0) {
            throw new IllegalArgumentException("Fuel must be a positive number");
        }
        if (fuelQuantity > tankCapacity) {
            throw new IllegalArgumentException("Cannot fit fuel in tank");
        }
        this.fuelQuantity = fuelQuantity;
    }

    protected void setFuelConsumption(double fuelConsumption) {
        this.fuelConsumption = fuelConsumption;
    }

    protected void setTankCapacity(double tankCapacity) {
        this.tankCapacity = tankCapacity;
    }

    @Override
    public String toString() {
        return String.format("%s: %.2f", this.getClass().getSimpleName(), this.fuelQuantity);
    }
}
